### PR TITLE
(feature) User flow

### PR DESCRIPTION
**Pull request by Blaumaus**
Ticket: https://github.com/Swetrix/roadmap/issues/57

### CODE REVIEW

File: src/analytics/dto/pageviews.dto.ts [hunk 1]
```diff
@@ -23,6 +23,12 @@ export class PageviewsDTO {
   })
   pg?: string
 
+  @ApiProperty({
+    example: '/home',
+    description: 'Previous page user was on',
+  })
+  prev?: string
```
Comment (yevheniikulisidi) [inline src/analytics/dto/pageviews.dto.ts:30]: Add validation!

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/analytics/analytics.controller.ts [hunk 2]
```diff
@@ -617,6 +442,46 @@ export class AnalyticsController {
...
+  ): Promise<any> {
+    const {
+      pid,
+      period,
+      timeBucket,
+      from,
+      to,
+      timezone = DEFAULT_TIMEZONE,
+    } = data
+    this.analyticsService.validatePID(pid)
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.controller.ts:471]: The validation must be in the DTO before the data reaches the controller.

File: src/analytics/analytics.controller.ts [hunk 3]
```diff
@@ -617,6 +442,46 @@ export class AnalyticsController {
...
+
+    if (!_isEmpty(period)) {
+      this.analyticsService.validatePeriod(period)
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.controller.ts:474]: Validate in the DTO.

File: src/analytics/analytics.controller.ts [hunk 4]
```diff
@@ -617,6 +442,46 @@ export class AnalyticsController {
...
+    }
+
+    this.analyticsService.validateTimebucket(timeBucket)
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.controller.ts:477]: Validate in the DTO.

File: src/analytics/analytics.service.ts [hunk 5]
```diff
@@ -341,6 +342,100 @@ export class AnalyticsService {
     return null
   }
 
+  getGroupFromTo(from: string, to: string, timeBucket: TimeBucketType, period: string, timezone: string): { groupFrom: string, groupTo: string } {
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.service.ts]: You can create a separate interface with response fields.

File: src/analytics/analytics.service.ts [hunk 6]
```diff
@@ -341,6 +342,100 @@ export class AnalyticsService {
...
+        throw new PreconditionFailedException(
+          "The timeframe 'from' parameter is invalid",
+        )
+      }
+
+      if (!isValidDate(to)) {
+        throw new PreconditionFailedException(
+          "The timeframe 'to' parameter is invalid",
+        )
+      }
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.service.ts:381]: Validate in the DTO.

File: src/analytics/dto/getUserFlow.dto.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  @ApiProperty()
+  to: string
+
+  @ApiProperty({
+    description: 'Timezone to display data in',
+    default: DEFAULT_TIMEZONE,
+  })
+  timezone: string
+}
```
Comment (yevheniikulisidi) [inline src/analytics/dto/getUserFlow.dto.ts:36]: Validate all fields!

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/analytics/analytics.service.ts [hunk 8]
```diff
@@ -132,6 +132,21 @@ interface GetFiltersQuery extends Array<string | object> {
...
+  source: string
+  target: string
+  value: number
+}
+
+interface IUserFlow {
+  nodes: IUserFlowNode[]
+  links: IUserFlowLink[]
+}
+
```
Comment (yevheniikulisidi) [inline src/analytics/analytics.service.ts:149]: Put each interface in a separate folder (e.g. `interface/`) in the files.

**Review by yevheniikulisidi — CHANGES_REQUESTED** [review]

File: src/analytics/dto/getUserFlow.dto.ts (hunk 7 above)
Comment (Blaumaus) [reply src/analytics/dto/getUserFlow.dto.ts:36]: this and other validations for analytics component should be done in a separate PR

**Pull request merged**